SI-7331, SI-6843, SI-7731, parser entry point refactoring

densh opened the pull request:
This commits fixes following bugs:
- SI-7331 tb.parse returns unpositioned trees
- SI-6843 better error reporting for quasiquotes
- SI-7731 CannotHaveAttrs has inconsistent handling of setPos and setType

And additionaly refactors parser entry points to make parsing stats consistent across repl, toolbox, script runner and quasiquotes.

densh committed:
refactor parser entry points and extract a few methods out

This commit contains three logical changes:

1. Split `templateStatSeq` into two methods as we need more reliable
   parsing of template body alone for new `parseStats` entry point.

2. Add new parser entry point called `parseStats` which is aimed towards
   use in tools that require parsing of Scala code that can be written
   inside of a template.

   Such functionality is required for parsing lines in repl, parsing
   code through toolbox, parsing and running scala scripts and lastly
   for quasiquotes. All of them are refactored to use this very method
   in the next commits.

   A new method called `templateStatsCompat` is also added to make this
   commit pass the tests but it's a temporary hack that will be removed
   in next commit in favor of `parseStats`.

3. Extract out a few methods like `isCaseDefStart`, `expectedMsgTemplate`
   and `parseRule`. These are needed to override parser behaviour in
   updated quasiquotes parser (see next commits).

densh committed:
refactor repl to use new new parser entry point

This commit refactors repl to use `parseStats` entry point and
streamlines hacky error handling that was previously used to
encode errors that happen during parsing.

densh committed:
SI-7731 make CannotHaveAttrs more consistent

Previously setPos, pos_=, setType, tpe_= all behaved inconsistently
between each other even though they all represent similar
attributes that cannot be changed on CannotHaveAttrs trees.

In order to simplify handling of such trees in compiler code each
of these fields now supports assignment to its current default
value: NoType for tpe and NoPosition for pos. Such assignments
don't mutate underlying trees.

densh committed:
SI-7331 remove all the wrapping code from toolbox

Change toolbox parsing to use `parseStats` parser entry point
instead of current code-wrappign technique that makes positions
much less useful to end users.

There is also no need to create a compiler `Run` for parsing.

densh committed:
SI-6843 well-positioned syntax errors for quasiquotes

This is achieved in a following way:

1. Similarly to toolbox quasiquotes can go away with wrapping for
   parsing purpose after introduction of `parseStats` and `parseRule`
   entry points.

2. In case of syntax error quasiquote computes equivalent corresponding
   position in the source code with the help of `corrrespondingPosition`
   mapper which relies on position data collected into `posMap` during
   code generation.

retronym commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -1017,14 +1017,16 @@ trait Trees extends api.Trees { self: SymbolTable =>
-    private def unsupported(what: String, args: Any*) =
-      throw new UnsupportedOperationException(s"$what($args) inapplicable for "+self.toString)
+    private def requireLegal(value: Any, allowed: Any, what: String) =
+      require(value == allowed, s"can't set $what for $self to value other than $allowed")
```
Could be unit tested. Probably fits best in a new JUnit test case under test/junit.

retronym commented:
s/two logical changes/three logical changes/?

densh committed:
testing utility function that lets to assert exceptions easily

This is much easier to use than built-in JUnit method-level checks.

densh commented:
PLS REBUILD/pr-scala@d5f4b5b

scala-jenkins commented:
(kitty-note-to-self: ignore 22395091)
:cat: Roger! Rebuilding pr-scala for d5f4b5b2. :rotating_light:

densh commented:
Updated pr with junit tests for `CannotHaveAttrs` and changes to `correspondingPosition` suggested by @retronym 

I've also added assertThrows utility function that needs additional review. Probably by @gkossakowski

scala-jenkins commented:
SPURIOUS ABORT? -- PLS REBUILD/pr-scala@d0c172a35eb5d729f613c9852fd959a2042c84c3

scala-jenkins commented:
SPURIOUS ABORT? -- PLS REBUILD/pr-scala@8647ef676d760cda059f3aea8c76f79fb85de7fa

scala-jenkins commented:
(kitty-note-to-self: ignore 22401704)
:cat: Roger! Rebuilding pr-scala for d0c172a3. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 22401705)
:cat: Roger! Rebuilding pr-scala for 8647ef67. :rotating_light:

densh commented:
Extended `CannotHaveAttrs` tests to include tests for all of the typical instances.

adriaanm commented:
PLS REBUILD/pr-scala@c22d877

scala-jenkins commented:
(kitty-note-to-self: ignore 22462541)
:cat: Roger! Rebuilding pr-scala for c22d8770. :rotating_light:
